### PR TITLE
Feature/656 target location

### PR DESCRIPTION
**Pull request by btry**

### CODE REVIEW

File: inc/targetbase.class.php [hunk 1]
```diff
@@ -551,7 +560,7 @@ protected function showUrgencySettings($rand) {
       ));
       $script = <<<EOS
          function change_urgency() {
-            $('#urgency_specific_question').hide();
+            $('#urgency_specific_title').hide();
```
Comment (btry) [inline inc/targetbase.class.php]: this is an UI fix found while developing the feature

File: inc/targetbase.class.php [hunk 2]
```diff
@@ -764,4 +773,76 @@ protected function parseTags($content, PluginFormcreatorForm_Answer $formanswer,
...
    }
+
+
+   protected function showLocationSettings($rand) {
+      global $DB;
+
+      echo '<tr class="line0">';
+      echo '<td width="15%">' . __('Location') . '</td>';
+      echo '<td width="45%">';
+      Dropdown::showFromArray('location_rule', static::getEnumLocationRule(), array(
```
Comment (orthagh) [inline inc/targetbase.class.php]: ``array(...)`` -> ``[...]``

File: inc/targetbase.class.php [hunk 3]
```diff
@@ -80,6 +80,15 @@ static function getEnumCategoryRule() {
       );
    }
 
+   static function getEnumLocationRule() {
+      return array(
```
Comment (orthagh) [inline inc/targetbase.class.php]: ``array(...)`` -> ``[...]``

File: inc/targetbase.class.php [hunk 4]
```diff
@@ -764,4 +773,76 @@ protected function parseTags($content, PluginFormcreatorForm_Answer $formanswer,
...
+
+      echo '<tr class="line0">';
+      echo '<td width="15%">' . __('Location') . '</td>';
+      echo '<td width="45%">';
+      Dropdown::showFromArray('location_rule', static::getEnumLocationRule(), array(
+         'value'                 => $this->fields['location_rule'],
+         'on_change'             => 'change_location()',
+         'rand'                  => $rand
+      ));
+      $script = <<<EOS
```
Comment (orthagh) [inline inc/targetbase.class.php]: it's much a personal preference but for javascript heredoc syntax, i use JAVASCRIPT tag name, thereby my editor (sublime or vscode) highlight correctly the code

File: inc/targetbase.class.php [hunk 5]
```diff
@@ -764,4 +773,76 @@ protected function parseTags($content, PluginFormcreatorForm_Answer $formanswer,
...
+
+      echo '<div id="location_specific_value" style="display: none">';
+      Location::dropdown(array(
+         'name' => '_location_specific',
+         'value' => $this->fields["location_question"],
+      ));
+      echo '</div>';
+      echo '<div id="location_question_value" style="display: none">';
+      // select all user questions (GLPI Object)
+      $query2 = "SELECT q.id, q.name, q.values
```
Comment (orthagh) [inline inc/targetbase.class.php]: ``DBmysqlIterator`` possible ?

File: inc/targetbase.class.php [hunk 6]
```diff
@@ -764,4 +773,76 @@ protected function parseTags($content, PluginFormcreatorForm_Answer $formanswer,
...
+      echo Html::scriptBlock($script);
+      echo '</td>';
+      echo '<td width="15%">';
+      echo '<span id="location_question_title" style="display: none">' . __('Question', 'formcreator') . '</span>';
+      echo '<span id="location_specific_title" style="display: none">' . __('Location ', 'formcreator') . '</span>';
+      echo '</td>';
+      echo '<td width="25%">';
+
+      echo '<div id="location_specific_value" style="display: none">';
+      Location::dropdown(array(
```
Comment (orthagh) [inline inc/targetbase.class.php]: ``array(...)`` -> ``[...]``

File: inc/targetbase.class.php [hunk 7]
```diff
@@ -764,4 +773,76 @@ protected function parseTags($content, PluginFormcreatorForm_Answer $formanswer,
...
+                AND q.fieldtype = 'dropdown'";
+      $result2 = $DB->query($query2);
+      $users_questions = array();
+      while ($question = $DB->fetch_array($result2)) {
+         $decodedValues = json_decode($question['values'], JSON_OBJECT_AS_ARRAY);
+         if (isset($decodedValues['itemtype']) && $decodedValues['itemtype'] === 'Location') {
+            $users_questions[$question['id']] = $question['name'];
+         }
+      }
+      Dropdown::showFromArray('_location_question', $users_questions, array(
```
Comment (orthagh) [inline inc/targetbase.class.php]: ``array(...)`` -> ``[...]``

File: install/update_2.5_2.6.php [hunk 8]
```diff
@@ -25,5 +25,23 @@ function plugin_formcreator_update_2_6(Migration $migration) {
...
+   // add location rule
+   $enum_location_rule = "'".implode("', '", array_keys(PluginFormcreatorTargetTicket::getEnumLocationRule()))."'";
+   if (!FieldExists('glpi_plugin_formcreator_targettickets', 'location_rule', false)) {
+      $query = "ALTER TABLE `glpi_plugin_formcreator_targettickets`
+                ADD `location_rule` ENUM($enum_location_rule) NOT NULL DEFAULT 'none' AFTER `category_question`;";
+      $DB->query($query) or plugin_formcreator_upgrade_error($migration);
+   } else {
+      $current_enum_location_rule = PluginFormcreatorCommon::getEnumValues('glpi_plugin_formcreator_targettickets', 'location_rule');
+      if (count($current_enum_location_rule) != count(PluginFormcreatorTargetTicket::getEnumLocationRule())) {
+         $query = "ALTER TABLE `glpi_plugin_formcreator_targettickets`
```
Comment (orthagh) [inline install/update_2.5_2.6.php]: use ``$migration->changeField``

File: install/update_2.5_2.6.php [hunk 9]
```diff
@@ -25,5 +25,23 @@ function plugin_formcreator_update_2_6(Migration $migration) {
       $DB->query($query) or plugin_formcreator_upgrade_error($migration);;
    }
 
+   // add location rule
+   $enum_location_rule = "'".implode("', '", array_keys(PluginFormcreatorTargetTicket::getEnumLocationRule()))."'";
+   if (!FieldExists('glpi_plugin_formcreator_targettickets', 'location_rule', false)) {
+      $query = "ALTER TABLE `glpi_plugin_formcreator_targettickets`
```
Comment (orthagh) [inline install/update_2.5_2.6.php]: use ``$migration->addField``

File: inc/targetbase.class.php (hunk 5 above)
Comment (btry) [reply inc/targetbase.class.php]: I planned to cleanup the code to 
- remove warnings raised by my IDE, 
- replace array() by [], 
- drop remaining raw SQL

one PR per file or small group of files

This code was done before I was sure to drop the compatibility with GLPI 9.1

File: install/update_2.5_2.6.php (hunk 9 above)
Comment (btry) [reply install/update_2.5_2.6.php]: As far as I know Migration does not know how to deal with ENUIM. That's why I keep this style for now.

If I'm wrong, please show me the trick

File: install/update_2.5_2.6.php (hunk 8 above)
Comment (btry) [reply install/update_2.5_2.6.php]: same answer as above

File: install/update_2.5_2.6.php (hunk 9 above)
Comment (orthagh) [reply install/update_2.5_2.6.php]: addField/changeField call Migration::fieldFormat which a list of shortcut cases

and a default case to copy directly the type (ENUM ... NOT NULL... etc)

Comment (btry) [reply install/update_2.5_2.6.php]: done

**Pull request merged**